research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+              batch_size=5, query_size=2000, neg_num=5, update_every=1,
+              debug=False):
+  """Executes either training or validation step based on `train` value.
+
+  Args:
+    loader: Training/validation iterable dataset.
+    model: Network to train/validate.
+    criterion: Loss function.
+    optimizer: Network optimizer.
+    epoch: Integer, epoch number.
+    train: Bool, specifies training or validation phase.
+    batch_size: Integer, number of (q,p,n1,...,nN) tuples in a mini-batch.
+    query_size: Integer, number of queries randomly drawn per one training
+      epoch.
+    neg_num: Integer, number of negatives per a tuple.
+    update_every: Integer, update model weights every N batches, used to
+      handle relatively large batches batch_size effectively becomes
+      update_every x batch_size.
+    debug: Bool, whether debug mode is used.
+
+  Returns:
+    average_epoch_loss: Average epoch loss.
+  """
+  batch_time = global_features_utils.AverageMeter()
+  data_time = global_features_utils.AverageMeter()
+  losses = global_features_utils.AverageMeter()
+
+  # Retrieve all trainable variables we defined in the graph.
+  tvs = model.trainable_variables
+  accum_grads = [tf.Variable(tf.zeros_like(tv.read_value()), trainable=False)
+                 for tv in tvs]
+
+  end = time.time()
+  batch_num = 0
+  print_frequency = 10
+  all_batch_num = query_size // batch_size
+  state = 'Train' if train else 'Val'
+  global_features_utils.debug_and_log('>> {} step:'.format(state))
+
+  # For every batch in the dataset; Stops when all batches in the dataset have
+  # been processed.
+  while True:
+    data_time.update(time.time() - end)
+
+    if train:
+      try:
+        # Train on one batch.
+        # We load batches into memory consequently.
+        for _ in range(batch_size):
+          # Because the images are not necessarily of the same size, we can't
+          # set the batch size with .batch().
+          batch = loader.get_next()
+          input_tuple = batch[0:-1]
+          target_tuple = batch[-1]
+
+          loss_value, grads = grad(criterion, model, input_tuple,
+                                   target_tuple, neg_num)
+          losses.update(loss_value)
+          # Adds to each element from the list you initialized earlier
+          # with zeros its gradient (works because accum_vars and gvs
+          # are in the same order).
+          accum_grads = [accum_grads[i].assign_add(gv) for i, gv in
+                         enumerate(grads)]
+
+        if (batch_num + 1) % update_every == 0 or (
+                batch_num + 1) == all_batch_num:
+          # Do one step for multiple batches. Accumulated gradients are
+          # used.
+          optimizer.apply_gradients(
+                  zip(accum_grads, model.trainable_variables))
+          accum_grads = [
+            tf.Variable(tf.zeros_like(tv.read_value()), trainable=False)
+            for tv in tvs]
+      except Exception as ex:
```

Comment:
We shouldn't use such catch-all exceptions. From the style guide: "Never use catch-all except: statements, or catch Exception or StandardError, unless you are" (https://google.github.io/styleguide/pyguide.html#244-decision)

I am wondering if this is really needed here, or if it could be removed. When do exceptions happen in this case?

Reply:
This is a stopping condition for loader.get_next()
We get a tensorflow.python.framework.errors_impl.OutOfRangeError: End of sequence
And then we know that we iterated through all of the pictures

Reply:
sounds good, thanks! Can you please add a comment to mention exactly this? Eg, something like "We break when we run out of range, ie, we exhausted all dataset images."